From f1ebb9e5f12a3ec55fa568c470ab57a47a910ccf Mon Sep 17 00:00:00 2001
From: mj <mj@maxlinear.com>
Date: Tue, 30 Sep 2025 23:35:00 -0700
Subject: [PATCH] WLANRTSYS-94977 Enable/Disable Zwdfs second best channel
 feature Feature: using dynamic_failsafe param to enable/disable zwdfs second
 best channel feature. dynamic_failsafe = 0; // fixed failsafe channel
 dynamic_failsafe = 1; // best non DFS channel dynamic_failsafe = 2; // zwdfs
 second best channel

---
 hostapd/mxl_config_proprietary.c |  3 ++-
 src/ap/mxl_common_proprietary.c  | 13 ++++++++-----
 src/ap/mxl_config.h              |  7 ++++++-
 src/ap/mxl_dfs_common.c          |  4 +++-
 src/ap/mxl_drv_callbacks.c       |  9 ++++++++-
 src/ap/mxl_hostapd.c             |  2 +-
 6 files changed, 28 insertions(+), 10 deletions(-)

diff --git a/hostapd/mxl_config_proprietary.c b/hostapd/mxl_config_proprietary.c
index 7ea7af1d2..dd6fbc1a5 100644
--- a/hostapd/mxl_config_proprietary.c
+++ b/hostapd/mxl_config_proprietary.c
@@ -522,7 +522,8 @@ memcpy_fail:
 continue_no_fail:
 #endif /* CONFIG_ACS */
 
-	mxl_conf->dynamic_failsafe = 1;
+	/* Enable ZWDFS second best channel feature */
+	mxl_conf->dynamic_failsafe = 2;
 	return 0;
 }
 
diff --git a/src/ap/mxl_common_proprietary.c b/src/ap/mxl_common_proprietary.c
index 193261cfd..aa5ed9b6c 100644
--- a/src/ap/mxl_common_proprietary.c
+++ b/src/ap/mxl_common_proprietary.c
@@ -137,6 +137,10 @@ mxl_dfs_get_channel_dynamic_failsafe(struct hostapd_iface *iface,
 		return NULL;
 	}
 
+	if (mxl_conf->dynamic_failsafe == dfs_failsafe_second_best_zwdfs) {
+		return channel;
+	}
+
 	cf1 = failsafe->freq + failsafe->width / 2 - 10;
 	ieee80211_freq_to_chan(cf1, vht_oper_centr_freq_seg0_idx);
 	*secondary_channel = mxl_hostapd_get_secondary_offset(channel->chan, failsafe->width, *vht_oper_centr_freq_seg0_idx);
@@ -156,11 +160,10 @@ mxl_dfs_get_channel_dynamic_failsafe(struct hostapd_iface *iface,
 			freq_params->ht_enabled = 1;
 		if (iface->conf->ieee80211ac)
 			freq_params->vht_enabled = 1;
-		if (mxl_conf->dynamic_failsafe == 0) {
-			mxl_hostapd_set_conf_from_freq_params(iface, freq_params);
-			*vht_oper_centr_freq_seg0_idx = hostapd_get_oper_centr_freq_seg0_idx(iface->conf);
-			*vht_oper_centr_freq_seg1_idx = hostapd_get_oper_centr_freq_seg1_idx(iface->conf);
-		}
+
+		mxl_hostapd_set_conf_from_freq_params(iface, freq_params);
+		*vht_oper_centr_freq_seg0_idx = hostapd_get_oper_centr_freq_seg0_idx(iface->conf);
+		*vht_oper_centr_freq_seg1_idx = hostapd_get_oper_centr_freq_seg1_idx(iface->conf);
 	}
 
 	return channel;
diff --git a/src/ap/mxl_config.h b/src/ap/mxl_config.h
index 28a979c93..b86d5531b 100644
--- a/src/ap/mxl_config.h
+++ b/src/ap/mxl_config.h
@@ -37,6 +37,12 @@ struct hostapd_bss_config;
 #define ACS_NOISE_THR_MAX               0
 #define ACS_NOISE_THR_DEFAULT           (-68)
 
+typedef enum {
+	dfs_failsafe_fixed_channel = 0,
+	dfs_failsafe_best_non_dfs,
+	dfs_failsafe_second_best_zwdfs
+} dfs_failsafe_modes;
+
 typedef struct acs_chan {
         int primary;
         int secondary;
@@ -372,7 +378,6 @@ struct mxl_hostapd_config {
 		ACS_ALGO_SURVEY = 0,
 		ACS_ALGO_SMART  = 1,
 	} acs_algo;
-
 #ifdef CONFIG_ACS
 	/* SmartACS */
 	int acs_init_done;
diff --git a/src/ap/mxl_dfs_common.c b/src/ap/mxl_dfs_common.c
index ed94dec5c..7a2f2fc1d 100644
--- a/src/ap/mxl_dfs_common.c
+++ b/src/ap/mxl_dfs_common.c
@@ -26,6 +26,7 @@
 #include "common/mxl_wpa_ctrl.h"
 #include "ap/mxl_hostapd.h"
 #include "mxl_acs.h"
+#include "mxl_config.h"
 
 /* Check if a 20 MHz channel number is DFS */
 int is_dfs_channel(int chan)
@@ -817,7 +818,8 @@ bool mxl_switch_to_sec_best_channel_wrap(struct hostapd_iface *iface, int chan_w
 #ifdef CONFIG_VENDOR_MXL_PROPRIETARY
 	struct mxl_hostapd_config *mxl_conf = &iface->conf->mxl_conf;
 
-	if (mxl_conf->dynamic_failsafe != 0) {
+	wpa_printf(MSG_INFO, "radar detected - zwdfs_second_best_ch_enable : %d", mxl_conf->dynamic_failsafe);
+	if (mxl_conf->dynamic_failsafe == dfs_failsafe_second_best_zwdfs) {
 		if (mxl_switch_second_best_channel(iface, chan_width)) {
 			return true;
 		}
diff --git a/src/ap/mxl_drv_callbacks.c b/src/ap/mxl_drv_callbacks.c
index 5fcf8f31c..ef7a334e4 100644
--- a/src/ap/mxl_drv_callbacks.c
+++ b/src/ap/mxl_drv_callbacks.c
@@ -24,6 +24,7 @@
 #include "ap_drv_ops.h"
 #include "beacon.h"
 #include "mxl_hostapd.h"
+#include "mxl_config.h"
 
 #ifdef CONFIG_IEEE80211BE_MXL_MLO
 void mxl_hostapd_notif_disassoc(struct hostapd_data *hapd, const u8 *addr)
@@ -235,9 +236,15 @@ void mxl_hostapd_event_eapol_rx(struct hostapd_data *hapd, const u8 *src,
 void mxl_check_ch_switch_status_and_get_second_best_channel(int finished, struct hostapd_data *hapd)
 {
 #ifdef CONFIG_VENDOR_MXL_PROPRIETARY
-	if(!hapd->iface->conf->mxl_conf.acs_init_done) {
+	struct mxl_hostapd_config *mxl_conf = &hapd->iconf->mxl_conf;
+
+	if (mxl_conf->dynamic_failsafe != dfs_failsafe_second_best_zwdfs)
+		return;
+
+	if (!hapd->iface->conf->mxl_conf.acs_init_done) {
 		return;
 	}
+
 	if (finished && !hapd->iface->mxl_iface.sec_best_ch.flag) {
 		if (WAVE_FREQ_IS_5G(hapd->iface->freq) && !(hapd->iface->mxl_iface.sec_best_ch.bg_cac)) {
 			if (mxl_is_dfs_channel(hapd->iface->conf->channel, hostapd_get_oper_chwidth(hapd->iface->conf)))
diff --git a/src/ap/mxl_hostapd.c b/src/ap/mxl_hostapd.c
index ed7c6eca6..479c2d4a2 100644
--- a/src/ap/mxl_hostapd.c
+++ b/src/ap/mxl_hostapd.c
@@ -1797,7 +1797,7 @@ void mxl_hostapd_setup_interface_post(struct hostapd_iface *iface)
 	int ap_retry_limit_cfg[3];
 
 #ifdef CONFIG_VENDOR_MXL_PROPRIETARY
-	if (WAVE_FREQ_IS_5G(iface->freq) && !(iface->mxl_iface.sec_best_ch.bg_cac)) {
+	if ((mxl_conf->dynamic_failsafe == dfs_failsafe_second_best_zwdfs) && WAVE_FREQ_IS_5G(iface->freq) && !(iface->mxl_iface.sec_best_ch.bg_cac)) {
 		if (mxl_is_dfs_channel(iface->conf->channel, hostapd_get_oper_chwidth(iface->conf)))
 		{
 			if (mxl_get_second_best_channel_from_acs_table(iface)) {
-- 
2.43.0

